File: cases/taint/csharp/mutual-recursive-transform-positive/MutualRecursiveTransformPositive.cs
namespace DataFlowBench;

static class MutualRecursiveTransformPositive
{
    static int dfb_source() // DFB-SOURCE: mutual-recursive-transform-input
    {
        return 7;
    }

    static void dfb_sink(int value) { } // DFB-SINK: mutual-recursive-transform-sink

    static int WalkA(int value, int depth)
    {
        if (depth == 0)
        {
            return value;
        }

        int next = value + 1;
        int recursive = WalkB(next, depth - 1); // DFB-WITNESS: mutual-recursive-transform-a-transfer
        return recursive + 1; // DFB-WITNESS: mutual-recursive-transform-a-compose
    }

    static int WalkB(int value, int depth)
    {
        if (depth == 0)
        {
            return value; // DFB-WITNESS: mutual-recursive-transform-b-base
        }

        int next = value + 1;
        int recursive = WalkA(next, depth - 1); // DFB-WITNESS: mutual-recursive-transform-b-transfer
        return recursive + 1; // DFB-WITNESS: mutual-recursive-transform-b-compose
    }

    static void Run()
    {
        dfb_sink(WalkA(dfb_source(), 3));
    }
}

File: cases/taint/kotlin/mutual-recursive-transform-positive/MutualRecursiveTransformPositive.kt
package dataflowbench

object MutualRecursiveTransformPositive {
    fun dfb_source(): Int { // DFB-SOURCE: mutual-recursive-transform-input
        return 7
    }

    fun dfb_sink(value: Int) {} // DFB-SINK: mutual-recursive-transform-sink

    fun walkA(value: Int, depth: Int): Int {
        if (depth == 0) {
            return value
        }
        val next = value + 1
        val recursive = walkB(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-a-transfer
        return recursive + 1 // DFB-WITNESS: mutual-recursive-transform-a-compose
    }

    fun walkB(value: Int, depth: Int): Int {
        if (depth == 0) {
            return value // DFB-WITNESS: mutual-recursive-transform-b-base
        }
        val next = value + 1
        val recursive = walkA(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-b-transfer
        return recursive + 1 // DFB-WITNESS: mutual-recursive-transform-b-compose
    }

    fun run() {
        dfb_sink(walkA(dfb_source(), 3))
    }
}
